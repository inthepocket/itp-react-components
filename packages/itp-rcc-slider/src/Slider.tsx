import { default as React, Fragment} from 'react';
import {
  Slider as RCSlider,
  Rail as RCRail,
  Handles as RCHandles,
  Tracks as RCTracks
} from 'react-compound-slider';
import { SliderProps } from 'react-compound-slider/Slider/Slider';
import Handle from './Handle';
import Track from './Track';

export interface SliderStylesInterface {
  Slider?: string;
  Rail?: string;
  Track?: string;
  Handle?: string;
  isHandleActive?: string;
}

export interface SliderPropsInterface extends Partial<SliderProps> {
  prefixCss?: string;
  styles?: SliderStylesInterface;
  left?: boolean;
  right?: boolean;
}

const Slider = ({
  left = false,
  prefixCss = 'slider',
  right = false,
  styles,
  values,
  ...sliderProps
}: SliderPropsInterface) => {
  const sliderClassName = styles ? styles.Slider : prefixCss;
  const railClassName = styles ? styles.Rail : `${prefixCss}__Rail`;

  return (
    <RCSlider
      className={sliderClassName}
      values={values}
      {...sliderProps}
    >
      <RCRail>
        {({ getRailProps }) => (
          <div className={railClassName} {...getRailProps()} />
        )}
      </RCRail>
      <RCHandles>
        {({ activeHandleID, handles, getHandleProps }) => (
          <Fragment>
            {handles.map(({ id, percent }) => (
              <Handle
                key={id}
                activeHandleID={activeHandleID}
                getHandleProps={getHandleProps}
                id={id}
                percent={percent}
                prefixCss={prefixCss}
                styles={styles}
              />
            ))}
          </Fragment>
        )}
      </RCHandles>
      <RCTracks left={left} right={right}>
        {({ tracks, getTrackProps }) => (
          <Fragment>
            {tracks.map(({ id, source, target }) => (
              <Track
                key={id}
                getTrackProps={getTrackProps}
                prefixCss={prefixCss}
                source={source}
                styles={styles}
                target={target}
              />
            ))}
          </Fragment>
        )}
      </RCTracks>
    </RCSlider>
  );
};

export default Slider;
