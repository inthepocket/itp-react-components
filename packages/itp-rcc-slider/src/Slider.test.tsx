import * as React from 'react';
import { create } from 'react-test-renderer';
import Slider from './Slider';

const cssModuleMock = {
  Handle: 'cssModuleHandle',
  Slider: 'cssModuleSlider',
  Track: 'cssModuleTrack',
  Rail: 'cssModuleRail',
  isHandleActive: 'cssModuleIsHandleActive',
};

describe('Slider', () => {
  test('single value', () => {
    const tree = create((
      <Slider
        prefixCss="Slider"
        values={[25]}
      />
    ))
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('single value, left track', () => {
    const tree = create((
      <Slider
        left={true}
        prefixCss="Slider"
        values={[25]}
      />
    ))
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('single value, right track', () => {
    const tree = create((
      <Slider
        right={true}
        prefixCss="Slider"
        values={[25]}
      />
    ))
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('single value, left + right track', () => {
    const tree = create((
      <Slider
        right={true}
        prefixCss="Slider"
        values={[25]}
      />
    ))
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('multiple values', () => {
    const tree = create((
      <Slider
        left={true}
        prefixCss="Slider"
        values={[25, 50, 75]}
      />
    ))
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('css modules', () => {
    const tree = create((
      <Slider
        left={true}
        values={[75]}
        styles={cssModuleMock}
      />
    ))
    .toJSON();

    expect(tree).toMatchSnapshot();
  });
});



