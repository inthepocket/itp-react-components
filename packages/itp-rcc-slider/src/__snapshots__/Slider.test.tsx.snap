// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider css modules 1`] = `
<div
  className="cssModuleSlider"
  style={Object {}}
>
  <div
    className="cssModuleRail"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  />
  <div
    className="cssModuleHandle"
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "75%",
      }
    }
  />
  <div
    className="cssModuleTrack"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "left": "0%",
        "width": "75%",
      }
    }
  />
</div>
`;

exports[`Slider multiple values 1`] = `
<div
  className="Slider"
  style={Object {}}
>
  <div
    className="Slider__Rail"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  />
  <div
    className="Slider__Handle"
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "25%",
      }
    }
  />
  <div
    className="Slider__Handle"
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "50%",
      }
    }
  />
  <div
    className="Slider__Handle"
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "75%",
      }
    }
  />
  <div
    className="Slider__Track"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "left": "0%",
        "width": "25%",
      }
    }
  />
  <div
    className="Slider__Track"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "left": "25%",
        "width": "25%",
      }
    }
  />
  <div
    className="Slider__Track"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "left": "50%",
        "width": "25%",
      }
    }
  />
</div>
`;

exports[`Slider single value 1`] = `
<div
  className="Slider"
  style={Object {}}
>
  <div
    className="Slider__Rail"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  />
  <div
    className="Slider__Handle"
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "25%",
      }
    }
  />
</div>
`;

exports[`Slider single value, left + right track 1`] = `
<div
  className="Slider"
  style={Object {}}
>
  <div
    className="Slider__Rail"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  />
  <div
    className="Slider__Handle"
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "25%",
      }
    }
  />
  <div
    className="Slider__Track"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "left": "25%",
        "width": "75%",
      }
    }
  />
</div>
`;

exports[`Slider single value, left track 1`] = `
<div
  className="Slider"
  style={Object {}}
>
  <div
    className="Slider__Rail"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  />
  <div
    className="Slider__Handle"
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "25%",
      }
    }
  />
  <div
    className="Slider__Track"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "left": "0%",
        "width": "25%",
      }
    }
  />
</div>
`;

exports[`Slider single value, right track 1`] = `
<div
  className="Slider"
  style={Object {}}
>
  <div
    className="Slider__Rail"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  />
  <div
    className="Slider__Handle"
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    role="slider"
    style={
      Object {
        "left": "25%",
      }
    }
  />
  <div
    className="Slider__Track"
    onMouseDown={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "left": "25%",
        "width": "75%",
      }
    }
  />
</div>
`;
